### PR TITLE
Update logging: bump lib and introduce pebble log forwarding

### PR DESCRIPTION
**Pull request by sed-i**
## Issue
- The charm is using an outdated loki_push_api charm lib.
- In air-gapped deployments, [promtail is not readily available for download](https://github.com/canonical/grafana-agent-k8s-operator/issues/300).

## Solution
- Bump loki_push_api to v1.
- Since air-gapped deployments are Juju 3.4+ anyway, use pebble log forwarding when possible.

Related to:
- https://github.com/canonical/grafana-agent-k8s-operator/issues/300
- https://github.com/canonical/operator/issues/1230

## Testing
Deploy the bundle:

```yaml
bundle: kubernetes
applications:
  loki:
    charm: loki-k8s
    channel: latest/edge
    base: ubuntu@20.04/stable
    resources:
      loki-image: 95
      node-exporter-image: 1
    scale: 1
    trust: true
  pg:
    charm: ./postgresql-k8s_ubuntu-22.04-amd64.charm
    resources:
      postgresql-image: ghcr.io/canonical/charmed-postgresql@sha256:31cf150b4523481202c1ff9b7b5d7f0b36729edad89d61242d8f1eb56b2912c0
    scale: 1
    trust: true
relations:
- - loki:logging
  - pg:logging
```

Confirm the pebble plan has a `log-targets` section:
```bash
juju exec --unit pg/0 -- PEBBLE_SOCKET=/charm/containers/postgresql/pebble.socket pebble plan
```
```yaml
# ...
log-targets:
    loki/0:
        type: loki
        location: http://loki-0.loki-endpoints.pg.svc.cluster.local:3100/loki/api/v1/push
        services:
            - all
        override: replace
        labels:
            charm: postgresql-k8s
            juju_application: pg
            juju_model: pg
            juju_model_uuid: 31887e0e-4c62-48df-85db-7f529bc25177
            juju_unit: pg/0
            product: Juju
```

Confirms logs are in loki:
```bash
LOKI=$(juju status --format=json | jq -r '.applications.loki.units."loki/0".address')

curl -sG $LOKI:3100/loki/api/v1/labels
# {"status":"success","data":["charm","juju_application","juju_model","juju_model_uuid","juju_unit","pebble_service","product"]}

curl -sG $LOKI:3100/loki/api/v1/label/juju_unit/values
# {"status":"success","data":["pg/0"]}

curl -sG $LOKI:3100/loki/api/v1/query_range --data-urlencode 'query={pebble_service="postgresql"}' | jq '.data.result[0]'
```
Trimmed down output:
```json
{
  "stream": {
    "charm": "postgresql-k8s",
    "juju_application": "pg",
    "juju_model": "pg",
    "juju_model_uuid": "31887e0e-4c62-48df-85db-7f529bc25177",
    "juju_unit": "pg/0",
    "pebble_service": "postgresql",
    "product": "Juju"
  },
  "values": [
    [
      "1716835072419000000",
      "server signaled"
    ],
    [
      "1716834778918000000",
      "2024-05-27 18:32:58 UTC [113]: user=,db=,app=,client=,line=2 HINT:  Future log output will appear in directory \"/var/log/postgresql\"."
    ],
    [
      "1716834777421000000",
      "performing post-bootstrap initialization ... ok"
    ],
    [
      "1716834777317000000",
      "The files belonging to this database system will be owned by user \"postgres\"."
    ]
  ]
}
```

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -173,17 +173,30 @@ def __init__(self, *args):
...
+            LogForwarder(
+                self,
+                relation_name="logging",
+            )
+            if self._is_pebble_log_forwarding_supported
+            else LogProxyConsumer(
+                self,
+                logs_scheme={"postgresql": {"log-files": POSTGRES_LOG_FILES}},
+                relation_name="logging",
+            )
```
Comment (sed-i) [inline src/charm.py:186]: Note that, while LogProxyConsumer is streaming everything from
```
POSTGRES_LOG_FILES = [
    "/var/log/pgbackrest/*",
    "/var/log/postgresql/patroni.log",
    "/var/log/postgresql/postgresql*.log",
]
```
using LogForwarder streams the stdout of all pebble services.

The two may differ. If critical, then could add a pebble service analogue to `tail -f` the files.

**Review by lucabello — APPROVED** [review]
Looks good to me!

File: src/charm.py (hunk 1 above)
Comment (phvalguima) [reply src/charm.py:186]: @marceloneppel @shayancanonical @dragomirp, wdyt?

Comment (sed-i) [reply src/charm.py:186]: It seems like there's a way config pg to log to stdout but I'm not clear on the syntax.
https://www.postgresql.org/docs/current/runtime-config-logging.html#GUC-LOGGING-COLLECTOR

Comment (dragomirp) [reply src/charm.py:186]: I don't think a tailing service or stdout will work for Pgbackrest out of the box. IIRC it creates separate log files for various commands. I don't see [an option](https://pgbackrest.org/configuration.html#section-log) to configure the logfile names. Maybe we can use `--log-level-console` and run the backup commands as a oneshot pebble service?

Comment (marceloneppel) [reply src/charm.py:186]: We probably can also use something like `--log-level-console` along with `container.exec([pgbackrest-command])wait_output()`.

**Comment by dragomirp** [comment]
Hi, @sed-i, sorry for the long delay. I've merged current main to the branch and created https://github.com/canonical/postgresql-k8s-operator/issues/627 and https://github.com/canonical/postgresql-k8s-operator/issues/628 to track the outstanding log forwarding issues. I think we can merge this.

**Review by dragomirp — APPROVED** [review]

**Review by marceloneppel — APPROVED** [review]

**Pull request merged**